### DISCUSSION TITLE
Point Stat Issue: Undefined Symbol

### DISCUSSION THREAD

**Discussion by wxninja** [dtcenter/METplus#1206] [Configuration] [answered]
Hi all-

I'm currently upgrading from MET 9.1 to MET 10.0, and have run into some trouble, apparently with my netCDF libraries? Maybe?

My build of 10.0.0 appears to have been successful--I'll include the config, install.log and make_test.log files showing the installation and test results.  I'll also include a snippet of my .cshrc file that governs the environment of installation.  You'll note that there are a few different netCDF library installation locations... as part of this process, I recompiled netCDF locally to make sure GCC versions matched and to try to help simplify any linker issues--but I don't think it worked.  

My full error is as follows when I run point_stat:

_/fewxops/software/met/met-10.0.0/bin/point_stat /fewxops/verification/data/model/grib2/feens1/2021/feens1_verify_2021101100f06.grib2 /fewxops/verification/data/obs/met_format/madis_meso/mesonet_fe_2021101106.nc /fewxops/software/met/met-10.0.0/data/config/PointStatConfig_default
DEBUG 1: Default Config File: /usr/local/share/met/config/PointStatConfig_default
DEBUG 1: User Config File: /fewxops/software/met/met-10.0.0/data/config/PointStatConfig_default
/fewxops/software/met/met-10.0.0/bin/point_stat: symbol lookup error: /fewxops/software/met/met-10.0.0/bin/point_stat: undefined symbol: _ZN6netCDF6NcFileC1ERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEENS0_8FileModeE_

There is no hint of this error in the make_test.log, so I'm a bit confused at the moment.

Two things to note: 1) the .nc file used here is a converted MADIS file, that was generated by madis2nc by the 9.1 version of MET. I'm checking to see if there was a RedHat upgrade in the past year (since 9.1 was installed) that could perhaps lead to some GCC version inconsistencies between 9.1 and 10.0.0.  Not sure if that would matter or not?

2) This error isn't specific to Point Stat.  I've also experienced it with madis2nc...and I'm guessing it's consistent with anything regarding netCDF. 

Any help with this is appreciated, as always! And please let me know if you need any additional information.
[configure.txt](https://github.com/dtcenter/METplus/files/7338178/configure.txt)
[cshrc.txt](https://github.com/dtcenter/METplus/files/7338179/cshrc.txt)
[make_install.log](https://github.com/dtcenter/METplus/files/7338180/make_install.log)
[make_test.log](https://github.com/dtcenter/METplus/files/7338181/make_test.log)

**Comment by jprestop** [comment, marked as answer]
Hi Tom (@wxninja). 

Searching through my emails, I found an [old MET-help ticket](https://mailman.ucar.edu/pipermail/met_help/2021-April/004387.html) from you with this same problem for a previous version (9.1.2) of MET.

I had said:

> Undefined symbol error messages like that are likely coming from the runtime linker.  Problems like that are usually solved by updating the LD_LIBRARY_PATH setting to tell the linker where to find the library archive files it needs at runtime.  Based on the error message, my guess is that the runtime linker isn't able to locate the NetCDF library.  If that is indeed the problem, you could try setting the LD_LIBRARY_PATH environment variable to the NetCDF directory which contains files named "libnetcdf.so*".  That might enable the linker to find that symbol.
> 
> Just FYI, in the future if you use the compile_MET_all.sh script to compile MET and its external libraries, it uses rpath, which designates the run-time search path hard-coded in an executable file, so this problem and the need to set LD_LIBRARY_PATH should be eliminated.

The compile_MET_all.sh script and its external libraries can be found on the [MET Download page](https://dtcenter.org/community-code/model-evaluation-tools-met/download) under "Sample Script For Compiling External Libraries And MET".  

You had said previously, like you mentioned here, that there were a few different netCDF library installation locations and said that you compiled MET with MET_NETCDFLIB pointing to both /usr/local/lib and /usr/local/external_libs/lib, but regardless of this, you kept getting the same error.  You also mentioned that you had rearranged the order of the libraries in LD_LIBRARY_PATH to account for the MET_NETCDFLIB first, then re-configured and re-built, but still no luck.  

Unfortunately, you left off with the following:

> I will rebuild the netcdf libraries with the newer version of gcc that is on this particular machine, which varies from the machine that the system files were mirrored over from.  I suspect this will help with the problem.
> I may not get to this until tomorrow, given I'm currently running a lot of archived data through MET 9.1 this morning and don’t want to disturb any environmental variables.  As soon as I recompile, I'll let you know it goes.

and I don't know how it worked out.

I am on my way out of the "virtual office" in a few minutes and I won't be back until Tuesday, but I will check back here for questions/updates from you and will reply as soon as I can.  **My suggestion is to use the compile_MET_all.sh script to compile MET, which will use "rpath" and will hopefully resolve the problems you are experiencing.** You don't have to reinstall the libraries that you already have built.  Instead, you'll just point to the needed MET variable locations in the configuration file you pass to the script.  You can find instructions for using the script on the Downloads page I linked to above, but also there are comments directly in the script that detail how to set the variables in your configuration file. You can also view these comments in our [GitHub repository here](https://raw.githubusercontent.com/dtcenter/MET/develop/scripts/installation/compile_MET_all.sh).  Please feel free to follow up with any questions.  I'll also try to check in before Tuesday if I am able.

**Comment by wxninja** [comment]
Thanks Julie!

I knew I ran into this issue with 9.1.2, but I apparently never solved it.  I believe I simply abandoned the 9.1.2 installation, knowing 10.0.0 was coming soon, and figured I'd pick it up from there--which I have, but still have the same issue. Kicking the can down the road is never the right answer, and I'm not sure why I thought it would be here.

I'm currently re-compiling zlib, hdf5 and netcdf libraries, as I do think there is a compiler version change somewhere in their update process over the last 12-18 months on this server. But I have a feeling that isn't the core issue.

I will certainly try the compile_MET_all.sh script--that seems the path of least resistance to getting 10.0.0 up and running.  I'm dying to know what is actually the cause of the error in the manual build, but ultimately a successful install is the most important outcome.  I'll try to get an install via compile_MET_all.sh  within the next day or two, and I'll report back on my progress.  I appreciate the offer, but there is no need to check in on my troubles when you're out of the office!

I appreciate your feedback, and I'll keep you posted.  Stay tuned!

**Comment by wxninja** [comment]
Hi Julie!

Success!  I used compile_MET_all.sh to build V10.0.0 (and all external libraries) in a separate local directory on the machine.  I'm working through testing now, but the build was successful and appears all functionality is running properly.  Thumbs up!

One thing that may (or may not!) be of interest:
I originally built via compile_MET_all.sh, but pointed to netCDF and HDF5 libraries (amongst others) that I had recently updated/rebuilt on the system (e.g. in /usr/local).  The build was successful, but netCDF funtions failed.  For example, madis2nc failed, as I believe it ran into the dreaded HDF5/netCDF cyclical dependency that has been known to show up when various installations exist on a system.  The error highlighted HDF5 headers being from 1.8.0, while the library was 1.12.0.  I would imagine this is related to my original trouble with the manual build--I must be crossing libraries that are inconsistent with their versions/compilers/builds somewhere.  When I pulled this back, and cut off all ties to previous libraries and existing MET environmental variables in order to build the entire package locally via the .sh script, we were successful.

In the end, this is a FAR better set up.  I really appreciate the effort the MET team has made in the compile_MET_all.sh script, as it will be extremely helpful for users to update MET going forward.  I had always built each version independently by hand, but ran the risk of the library inconsistencies when pointing to system libraries/settings.  Keeping MET  in its own package and directory structure will be far easier going forward.

If I get time in the future, I may try to build from hand again, just to see if I can better understand where the original problem came from here....but we'll see if I ever get there!

As always, thanks for you help (and patience).

**Reply by jprestop** [reply]
Thanks for the follow up to let us know that you were able to successfully compile MET and its external libraries using compile_MET_all.sh.  It is tricky when there are multiple installations of libraries on systems. If you do end up compiling manually again and are able to better understand where the original problem came from, I'd love for you to follow up with anything you learn.  I'm always eager to learn more about working through problems like those.  Thanks for the kind words about the script - we're glad it worked well for you.  At this point I'm going to mark this conversation as answered and lock it to prevent future posts. Our team has decided that once discussions have been answered, and the answer has been confirmed, we'll lock them. We want to encourage users to ask new questions in new discussions rather than posting to old ones. Hopefully that'll make the questions/answers easier for other users to follow. So if/when more issues/questions arise, please feel free to start a new discussion.  If you do end up wanting to follow up on this ticket with more information please go ahead and post to a new Discussion and reference this ticket in your Discussion. Thanks!
